### ISSUE TITLE
AJAX: errors are silently ignored by default, only logs are printed to the console

### ISSUE THREAD

**Issue by lprimak** [primefaces/primefaces#14773]
### Describe the bug

By default, all Ajax errors are silently ignored.
This is especially bad when dealing with unreliable networks, such as mobile phones.

I foud the offending code: https://github.com/primefaces/primefaces/blob/cfc7e5091379316e68e50b32ced90fb5223cd5d5/primefaces/src/main/frontend/packages/core/src/core/core.ajax.ts#L934

1. Error 401 rarely, if ever includes the "Location" header. Since it rarely exists, no redirection happens. In that case, instead of ignoring it, a redirection to self should occur, which will handle the logic correctly.

2. It only calls onerror if exists in spefic element, or if ajaxStatus component exists. There is no other way to intercept the error. ajaxStatus is too heavy for my needs.
I believe there needs to be a global error handler, like the one is jakarta.faces.js so a global hook can be made into the error chain.

3/ By default, if no error handlers are spefied, redirect-to-self should be done, instead of just pring to the console. This way, default functionality will be correct and not just hang the application

Summary: if no better outcome is possible, redirect to self for any ajax errors

JSF issue: https://github.com/jakartaee/faces/issues/2126

Discord thread link:
https://discord.com/channels/787967399105134612/1483207925731164343

### PrimeFaces edition

Community

### PrimeFaces version

Latest

### Theme

Any

### JSF implementation

Mojarra

### JSF version

Latest

### Java version

Latest

### Browser(s)

Any

**Comment by melloware** [comment]
Makes sense. Could you submit a PR suggesting your changes so we can review?

**Comment by tandraschko** [comment]
Im not sure if redirect to itsself is a real solution; just a hack in a undefined situation.
IMO a application either need error pages, p:ajaxStatus, global p:ajaxExceptionHandler or you intercept pfAjaxError.

for 2), there actually is global listener -> https://github.com/primefaces/primefaces/blob/cfc7e5091379316e68e50b32ced90fb5223cd5d5/primefaces/src/main/frontend/packages/core/src/core/core.ajax.ts#L950C17-L950C78

**Comment by tandraschko** [comment]
If you wanna do a redirect to self, you can just add it by your own while intercepting pfAjaxError.
Feel free to add it to the docs -> https://primefaces.github.io/primefaces/15_0_0/#/core/errorhandling

**Comment by lprimak** [comment]
Thomas,

The default behavior currently is not sufficient. The question is what to do by default in production mode.
Currently, the behavior is not good, and is supported by actual users on mobile phones complaining.
(`p:commandButton`) just remains greyed out when pressed, with no solution but manually reload the page.

Question is, how to improve it? I have been using redirect-to-self for many years, and the users are happy with it.
Ajax behaves well with it, due to stateful nature of the apps. Users intuitively understand what's happening.

I believe is that, in absence of an error handler, a good default behavior is just to reload the page.

What are the alternatives?
* Current (buttons remain greyed out) - not a good default IMHO
* Alert - Not good in prod mode, but good in dev mode
* Console error - Insufficient because it can't be seen on a mobile phone
Any others that I am missing?

**Comment by melloware** [comment]
@lprimak a reproducer showing the issue would be great I can simply turn off the network in CHrome and reproduce it.

I would like to see the CommandButton being greyed out and going nowhere.

**Comment by lprimak** [comment]
Showcase can be used as a reproducer here.

1. Load showcase
2. Turn on airplane mode and turn off wifi on your phone
3. just use any of "Ajax Disabling and Status Icon" buttons in the showcase

**Comment by tandraschko** [comment]
Refresh the page without a hint for the user is not a good way from ux perspective and will never be our default.
Adding as possibility and document it is the right way. Maxbe we can setup a small page for "first steps with pf, Whats important". Setup Error Handling is one of it

**Comment by lprimak** [comment]
Thomas, do you really think that disabled buttons with spinners hanging is better than a page reload, which will, show the user that their connection is down?
This was very well tested in many scenarios with actual users for many, many years.

Of course, the user can register their own handler. This is all about the default behavior, which is bad currently.

**Comment by melloware** [comment]
@tandraschko yes the use case he is talking about above should be handled.  Just turning on WIFI connectivity currently does not look like its being handled correctly to me?

**Comment by tandraschko** [comment]
Even that connection-error must trigger onerror callbacks/ajaxStatus/etc

We could add:
If Error=connection loss, then refresh

But thats the only case it makes sense 
Even then, If the page refreshes sucessful and is available again, the user is confused
Its just a better behavior when the refresh triggers really  again a "http connection error"

For every other case, a user have to configure error handling. The best solution would be that PF shows a styled error dialog with real messages.... But i18n and so on, just out of scope

**Comment by lprimak** [comment]
Yes, that would be better, but still incomplete IMHO.

By default, the app should never "hang" Reload would reveal the actual error, or yes, it would reload the page. It may go back to a login screen, or a transient bug / error would be bypassed.

Better than "hanging" disabled button and an inconsistent screen, and all the errors are either evident to the user, or evident to the back-end to be able to debug.

**Comment by lprimak** [comment]
This issue comes up not only on connection errors, but on spurious 404s, 500s, 504s, 401s, 403s etc. that come up due to servers going up/down, gateways timeouts, and things that happen in multi-node production environments, reverse proxys, k8s, etc.

**Comment by tandraschko** [comment]
Sorry but always refreshing the page is for me just another incomplete error handling.
- Loss of user control
- No explanation of what went wrong
- Confusing UX

As i stated above, even if its just a "connection error", still refreshing the page has this drawbacks above, if the connection is already up again on refreshing the page.

I dont say that the current behavior is good, its really bad, but as long we dont have a good solution for all cases, we wont change the default. IMO the only solution is to configure error handling in your application. This is up to your application and you can configure this in your global template.

However, we have some points here however we can work on:

1) document, that you need to setup error handling for a real-world application, in our getting started
2) We should allow to make a "refresh the page on-error" possible with a global hook. i think this is already possible, we should proove and document it
3) Verify that our error-callbacks are called on connection errors (doesnt matter if pfAjaxError, ajaxStatus, ajaxExceptionHandler)
4) Verify that we have error handling configured in our showcase

**Comment by lprimak** [comment]
> Sorry but always refreshing the page is for me just another incomplete error handling.

> * Loss of user control
> * No explanation of what went wrong
> * Confusing UX

I am not sure I understand your concerns here...

How is refreshing the page by default any more "loss of user control" vs. greyed out buttons and inconsistent state? I would say refresh is better, or, in the worst-case scenario, the same.
How is greyed out buttons / inconsistent state explains what went wrong any better than refreshing the page?
How is greyed out buttons / inconsistent state is any less confusing than refreshing the page?

> document, that you need to setup error handling for a real-world application, in our getting started
We should allow to make a "refresh the page on-error" possible with a global hook. i think this is already possible, we should proove and document it
Verify that our error-callbacks are called on connection errors (doesnt matter if pfAjaxError, ajaxStatus, ajaxExceptionHandler)
Verify that we have error handling configured in our showcase

* Points 2-3 are verified by me, and are definitely possible. The issue is that this should be default.
* Point 4 - showcase is not configured to handle errors, the issues can be reproduced readily with showcase.

**Comment by tandraschko** [comment]
Switching from "no error handling" to another incomplete error handling is not something I would push to a framework like PrimeFaces, sorry.

To be honest, it would be valid to add `alert("You don't have error handling configured, please contact the admin")` as a default - that would at least be transparent about whats happening. 

PrimeFaces is a component library. We provide building blocks, not a fully configured application. Just like you need to add your own header, footer, navigation, or security, error handling is something that needs to be configured by the application developer.

Even the Faces impls themselves dont provide a default error handling. Its a complex topic once you consider all the things an application needs to decide: styling, i18n, error page locations, retry strategies, logging, user notifications, and so on. There is no one-size-fits-all solution here.

**Comment by lprimak** [comment]
Thomas,
you are correct that my proposed solution is not perfect, but it's clearly better than the status quo.
However, I can see clearly that you are not willing to accept my solution, so I will no longer advocate for it.
I am going to create another issue, for buttons specifically.

**Comment by tandraschko** [comment]
Closing in favor of the follow up tickets.
